concordium-consensus/src/Concordium/GlobalState.hs

```diff
@@ -546,21 +553,57 @@ instance GlobalStateConfig DiskTreeDiskBlockConfig where
     type GSState DiskTreeDiskBlockConfig pv = SkovPersistentData pv (HashedPersistentBlockState pv)
     type GSContext DiskTreeDiskBlockConfig pv = PersistentBlockStateContext pv
 
-    initialiseExistingGlobalState _ DTDBConfig{..} = do
+    initialiseExistingGlobalState (_ :: SProtocolVersion pv) pvgd DTDBConfig{..} = do
       -- check if all the necessary database files exist
-      existingDB <- checkExistingDatabase dtdbTreeStateDirectory dtdbBlockStateFile
-      pbscCache <- liftIO $ Accounts.newAccountCache (rpAccountsCacheSize dtdbRuntimeParameters)
-      if existingDB then do
-        pbscBlobStore <- liftIO $ do
-          -- the block state file exists, is readable and writable
-          -- we ignore the given block state parameter in such a case.
-          loadBlobStore dtdbBlockStateFile
+      let treeStateFile = dtdbTreeStateDirectory </> "data.mdb"
+      bsPathEx <- liftIO $ doesPathExist dtdbBlockStateFile
+      tsPathEx <- liftIO $ doesPathExist treeStateFile
+      if bsPathEx then do
+        checkRWFile dtdbBlockStateFile BlockStatePermissionError
+        pbscBlobStore <- liftIO $ loadBlobStore dtdbBlockStateFile
+        pbscCache <- liftIO $ Accounts.newAccountCache (rpAccountsCacheSize dtdbRuntimeParameters)
         let pbsc = PersistentBlockStateContext{..}
-        logm <- ask
-        skovData <- liftIO (runLoggerT (loadSkovPersistentData dtdbRuntimeParameters dtdbTreeStateDirectory pbsc) logm
-                    `onException` (closeBlobStore pbscBlobStore))
-        return (Just (pbsc, skovData))
-      else return Nothing
+        if tsPathEx then do
+          checkRWFile treeStateFile TreeStatePermissionError
+          mapM_ (logEvent TreeState LLTrace) ["Existing database found.", "TreeState filepath: " ++ show dtdbBlockStateFile, "BlockState filepath: " ++ show treeStateFile]
+          logm <- ask
+          skovData <- liftIO $ runLoggerT (loadSkovPersistentData dtdbRuntimeParameters dtdbTreeStateDirectory pbsc) logm
+                                   `onException` closeBlobStore pbscBlobStore
+          return (Just (pbsc, skovData))
+        else do
+          logEvent GlobalState LLTrace "Block state file exists, but tree state file does not. Attempting to recreate the tree state starting from the genesis block."
+          let reGS = do
+                let genBSRef :: BlobRef (BlockStatePointers pv) = BlobRef 0
```

Comment:
I'm not sure how this is supposed to work. `BlobRef 0` is almost certainly not a valid `BlobRef (BlockStatePointers pv)`, but it seems to be used as the genesis block state.